### DISCUSSION TITLE
FORTRAN API for Loop/Iteration information

### DISCUSSION THREAD

**Discussion by thedrbubba** [LLNL/Caliper#371] [Q&A]
Hi,

I am learning how to use Caliper in Fortran. I may have overlooked it but I am looking for
the Fortran API to handle loop/iterations like in C++ as follows:

_CALI_CXX_MARK_LOOP_BEGIN
CALI_CXX_MARK_LOOP_ITERATION
CALI_CXX_MARK_LOOP_END_

I have looked at the examples and at the Fortran API:
        _https://github.com/LLNL/Caliper/blob/master/src/interface/c_fortran/wrapfcaliper.F_

I have setup the fortran examples and tried to use the loop-report hoping for iteration information but I am not getting iteration information out. 

Have I overlooked something? Will I have to just adapt what Fortran calls available to do what I need?

Thanks in advance for your time,
Terry

**Comment by daboehme** [comment]
Hi @thedrbubba ,

It is possible to do loop annotations in Fortran, but it is a bit more awkward as we lack the high-level macros. You'll first have to find Caliper's built-in "loop" attribute, and create a loop iteration attribute for your loop with `cali_make_loop_iteration_attribute()`. Then, use the `cali_begin_string()` and `cali_begin_int()` subroutines to start the loop and iteration regions, respectively, and end them with `cali_end()`. Here's an example:

```
    use caliper_mod
    use iso_c_binding, ONLY : C_INT64_T
    integer(C_INT64_T) : loop_attribute, iter_attribute

    loop_attribute = cali_find_attribute('loop')
    iter_attribute = cali_make_loop_iteration_attribute('mainloop')

    call cali_begin_string(loop_attribute, 'mainloop')

    do i = 1, 4
        call cali_begin_int(iter_attribute, i)

        call cali_end(iter_attribute)
    end do

    call cali_end(loop_attribute)
```

I've updated the `fortran-example` program to do that. Now you should be able to use `loop-report` and see something like this:

```
$ ./examples/apps/fortran-example loop-report,iteration_interval=1

Loop summary:
------------

Loop     Iterations Time (s) Iter/s (min) Iter/s (max)  Iter/s (avg) 
mainloop          4 0.000072     0.000000 166666.666667 89610.389610 

Iteration summary (mainloop):
-----------------

Block Iterations Time (s) Iter/s        
               0 0.000007      0.000000 
    1          1 0.000042  23809.523810 
    2          1 0.000011  90909.090909 
    3          1 0.000006 166666.666667 
    4          1 0.000006 166666.666667
```

Hope this helps!

**Reply by thedrbubba** [reply]
Thank you! I will "play" with it and see if it will do what I need. I appreciate your time to provide.
